sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala

```diff
@@ -157,64 +154,100 @@ case class InnerOuterEstimation(join: Join) extends Logging {
+      val lInterval = ValueInterval(leftKeyStat.min, leftKeyStat.max, leftKey.dataType)
+      val rInterval = ValueInterval(rightKeyStat.min, rightKeyStat.max, rightKey.dataType)
       if (ValueInterval.isIntersected(lInterval, rInterval)) {
-        // Get the largest ndv among pairs of join keys
-        val maxNdv = leftKeyStats.distinctCount.max(rightKeyStats.distinctCount)
-        if (maxNdv > ndvDenom) ndvDenom = maxNdv
+        val (newMin, newMax) = ValueInterval.intersect(lInterval, rInterval, leftKey.dataType)
+        val card = joinCardByNdv(leftKey, rightKey, newMin, newMax)
+        // Return cardinality estimated from the most selective join keys.
+        if (card < minCard) minCard = card
       } else {
-        // Set ndvDenom to zero to indicate that this join should have no output
-        ndvDenom = 0
+        // One of the join key pairs is disjoint, thus the two sides of join is disjoint.
+        minCard = 0
       }
       i += 1
     }
+    minCard
+  }
 
-    if (ndvDenom < 0) {
-      // We can't find any join key pairs with column stats, estimate it as cartesian join.
-      1
-    } else if (ndvDenom == 0) {
-      // One of the join key pairs is disjoint, thus the two sides of join is disjoint.
-      0
-    } else {
-      1 / BigDecimal(ndvDenom)
+  /** Compute join cardinality using the basic formula, and update column stats for join keys. */
+  private def joinCardByNdv(
+      leftKey: AttributeReference,
+      rightKey: AttributeReference,
+      newMin: Option[Any],
+      newMax: Option[Any]): BigInt = {
+    val leftKeyStat = leftStats.attributeStats(leftKey)
+    val rightKeyStat = rightStats.attributeStats(rightKey)
+    val maxNdv = leftKeyStat.distinctCount.max(rightKeyStat.distinctCount)
+    // Compute cardinality by the basic formula.
+    val card = BigDecimal(leftStats.rowCount.get * rightStats.rowCount.get) / BigDecimal(maxNdv)
+
+    // Update intersected column stats.
+    val newNdv = leftKeyStat.distinctCount.min(rightKeyStat.distinctCount)
+    val newMaxLen = math.min(leftKeyStat.maxLen, rightKeyStat.maxLen)
+    val newAvgLen = (leftKeyStat.avgLen + rightKeyStat.avgLen) / 2
+
+    join.joinType match {
+      case LeftOuter =>
+        keyStatsAfterJoin.put(leftKey, leftKeyStat)
+        keyStatsAfterJoin.put(rightKey,
+          ColumnStat(newNdv, newMin, newMax, rightKeyStat.nullCount, newAvgLen, newMaxLen))
+      case RightOuter =>
+        keyStatsAfterJoin.put(leftKey,
+          ColumnStat(newNdv, newMin, newMax, leftKeyStat.nullCount, newAvgLen, newMaxLen))
+        keyStatsAfterJoin.put(rightKey, rightKeyStat)
+      case FullOuter =>
+        keyStatsAfterJoin.put(leftKey, leftKeyStat)
+        keyStatsAfterJoin.put(rightKey, rightKeyStat)
+      case _ =>
+        val newStats = ColumnStat(newNdv, newMin, newMax, 0, newAvgLen, newMaxLen)
+        keyStatsAfterJoin.put(leftKey, newStats)
+        keyStatsAfterJoin.put(rightKey, newStats)
```

Comment:
The above code changes are new, right?

Reply:
They are a part of the old method [getIntersectedStats](https://github.com/apache/spark/pull/19531/files#diff-6387e7aaeb7d8e0cb1457b9d0fe5cd00L235). I need to revert those lines for outer join cases, thanks for pointing it out.